qiskit/algorithms/evolvers/imaginary/imaginary_evolver.py

```diff
@@ -0,0 +1,21 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base class for Quantum Imaginary Time Evolution used for typing purposes."""
+
+from abc import ABC
+
+from ..evolver import Evolver
+
+
+class ImaginaryEvolver(Evolver, ABC):
```

Comment:
Why do we need this? We should introduce a subclass like this if there is no specialization. Just labeling this class name imaginary and the other real doesn't actually enforce that. We should add extra layers to the class hierarchy like this unless they're specializing anything from the parent class. If that's just missing from this iteration I would suggest waiting to add this until we are doing something different with it.

Reply:
Hi,
So we do not have plans to specialize it. I agree that it is not what good practices would recommend. The reasoning was the following.

We will have algorithms that implement real evolution and algorithms that implement imaginary one. These time evolution algorithms will sometimes power other algorithms. 

For example, Gibbs state preparation algorithm can only use imaginary time evolution. By introducing ImaginaryEvolver, we wanted to have a handle that would cover all imaginary algorithms (they would extend this class). It would help, e.g. in the Gibbs state preparation, to enforce that a time evolution algorithm provided indeed runs imaginary time evolution (by checking the type). Otherwise, we would throw an error.

Another way of handling this would be to get rid of these inelegant classes and introduce a field or a method to each time evolution algorithm that would state whether it is real or imaginary. This can also be considered inelegant in a different way.

If you have a suggestion and additional arguments I would be happy to hear them. I might consider changing the approach here.

Reply:
I would probably remove the common evolver base class, and make each one of these abstract classes independent. They can have the same `evolve()` method signature (but it's fairly straightforward) and you can put different docstrings in each saying they evolve `exp(i tH)` vs `exp(tH)`. This would be what I'd suggest if you never anything to implement the base class directly or expect to take an `Evolver` class as an input type.

If that doesn't work I'd suggest dropping the subclass and adding a EvolverType enum that defines `real` and `imaginary` (and maybe `both`) as entries in it. Then have an abstract `evolver_type` attribute in the `Evolver` class that returns one of these to say what type of evolver it is. But you should only do this if you expect algorithms to work with either real or imaginary (which according to @Cryoris shouldn't be a thing).